src/cultionet/models/lightning.py

```diff
@@ -909,10 +912,40 @@ def validation_step(self, batch: Data, batch_idx: int = None) -> dict:
         }
         if 'crop_type_f1' in eval_metrics:
             metrics['vctf1'] = eval_metrics['crop_type_f1']
+
         self.log_dict(metrics, on_step=False, on_epoch=True, prog_bar=True)
 
+        if self.save_batch_val_metrics:
+            self._save_batch_metrics(metrics, self.current_epoch, batch)
+
         return metrics
 
+    def _save_batch_metrics(
+        self,
+        metrics: T.Dict[str, torch.Tensor],
+        epoch: int,
+        batch: Data
+    ) -> None:
+        """Saves batch metrics to a parquet file
+        """
+        if not self.trainer.sanity_checking:
+            write_metrics = {
+                'epoch': [epoch] * len(batch.train_id),
```

Comment:
interesting, so the epoch here is multiple by the length of all the training data, including the augmentation?

Reply:
Not all of the training data, just the training data in the batch. E.g., each batch (if batch size = 8) will have 8 different training ids, which could be any of the data--raw or augmented.

Because metrics are aggregated for each batch, each data id in a batch will have the same metric as all data ids in the batch. However, data sampling is random, so a data id will not likely have the same combination of batch ids each epoch. And because data batches are grouped, this is about as detailed as we can get without becoming inefficient.

Reply:
Thanks for the explanation Jordan 👍 
The log with each batch id is very detailed 👍